### PR TITLE
[WIP] Provide an easy way to write documentation

### PR DESCRIPTION
**Pull request by MangelMaxime**
This PR aims to provide an easy way to write documentation.

It's use [Fornax](https://gitlab.com/Krzysztof-Cieslak/Fornax) which is a static site generator define via F#. The DSL is really close to the one use in Fable.React.

Here is a GIF showing the current layout status and also how easy it is to add a new page. We can customize almost anything as Fornax use F# fsi to generate the website.

fable_powerpack_fornax_demo

PS: As always, I use Bulma for the CSS framework :smiley:

### CODE REVIEW

**Comment by MangelMaxime** [comment]
Ok, so I made some really basic documentation if someone can review them that would be great :)

I know how to use the lib but writing clean and concise documentation is another thing ^^.

Also, @Pauan I must be missing something about how to use `Json` module. If you look at the code I made for example, the benefit of it isn't obvious IMO if you could take a look would be awesome :)

To build the documentation, you need to have [Fornax](https://gitlab.com/Krzysztof-Cieslak/Fornax) installed and also the [Sass compiler](http://sass-lang.com/install).

Before merging, I still need to automate the publish of the docs site via Fake.

**Comment by M0ns1gn0r** [comment]
Thanks for the docs!

Would be also cool to have a sample of a non-JSON fetch request. E.g. a POST to upload a file.

**Comment by Pauan** [comment]
@MangelMaxime I would write that example like this:

```fsharp
let unwrapNumber a =
    match a with
    | Json.Number a -> a
    | _ -> failwith "Invalid JSON, it must be a number"

let unwrapString a =
    match a with
    | Json.String a -> a
    | _ -> failwith "Invalid JSON, it must be a string"

let unwrapObject a =
    match a with
    | Json.Object a -> Map.ofArray a
    | _ -> failwith "Invalid JSON, it must be an object"

let unwrapArray a =
    match a with
    | Json.Array a -> a
    | _ -> failwith "Invalid JSON, it must be an array"
```

```fsharp
open Fable.PowerPack
open Fable.PowerPack.Result


type UserDef =
    { firstname : string
      country : string }

type JsonDef =
    { size : int
      users : UserDef list }


let jsonTxt = 
    """
{
    size: 3
    users: [
        {
            firstname: "Maxime",
            country: "France"
        },
        {
            firstname: "Alfonso",
            country: "Spain"
        },
        {
            firstname: "Forki",
            country: "Germany"
        }
    ]
}        
    """


let json = jsonTxt |> Json.ofString |> unwrapResult |> unwrapObject

let sizeValue = json |> Map.find "size" |> unwrapNumber
let users = json |> Map.find "users" |> unwrapArray

let userList =
    [ for user in users do
        let obj = user |> unwrapObject
        let firstname = obj |> Map.find "firstname" |> unwrapString
        let country = obj |> Map.find "country" |> unwrapString
        yield { firstname = firstname
                country = country }
    ]

{ size = int sizeValue
  users = userList }
```

The first code is utility functions which should probably be in `fable-powerpack`, and the second code is the actual user code.

The benefit of the `Json` module is that it uses ADTs, so it forces you to correctly handle every possibility.

Because of that you have to use functions like `unwrapString`. This guarantees that it will always throw exceptions if the JSON is wrong.

The above code uses exceptions, but you can also use `Result` for even more type safety:

```fsharp
let number a =
    match a with
    | Json.Number a -> Ok(a)
    | _ -> Error(System.Exception("Invalid JSON, it must be a number"))

let string a =
    match a with
    | Json.String a -> Ok(a)
    | _ -> Error(System.Exception("Invalid JSON, it must be a string"))

let lookup (key: string) (a: Map<string, Json.Json>) =
    match Map.tryFind key a with
    | Some(a) -> Ok(a)
    | None -> Error(System.Exception("Could not find key " + key))

let object a =
    match a with
    | Json.Object a -> Ok(Map.ofArray a)
    | _ -> Error(System.Exception("Invalid JSON, it must be an object"))

let array a =
    match a with
    | Json.Array a -> Ok(a)
    | _ -> Error(System.Exception("Invalid JSON, it must be an array"))
```

```fsharp
open Result
open Fable.PowerPack
open Fable.PowerPack.Result


type UserDef =
    { firstname : string
      country : string }

type JsonDef =
    { size : int
      users : UserDef list }


let jsonTxt = 
    """
{
    size: 3
    users: [
        {
            firstname: "Maxime",
            country: "France"
        },
        {
            firstname: "Alfonso",
            country: "Spain"
        },
        {
            firstname: "Forki",
            country: "Germany"
        }
    ]
}        
    """


result {
    let! json = jsonTxt |> Json.ofString |> bind object

    let! sizeValue = json |> lookup "size" |> bind number
    let! users = json |> lookup "users" |> bind array

    let! userList =
        seq {
            for user in users ->
                result {
                    let! obj = user |> object
                    let! firstname = obj |> lookup "firstname" |> bind string
                    let! country = obj |> lookup "country" |> bind string
                    return { firstname = firstname
                             country = country }
                }
        }

    return { size = int sizeValue
             users = List.ofSeq userList }
}
```

(The above code doesn't work yet, it requires some changes to the `Fable.PowerPack.Result` module first)

Of course you can use pattern matching as well, if you want to handle multiple different types:

```fsharp
match a with
| Json.Number a -> ...
| Json.String a -> ...
| Json.Array a -> ...
| _ -> failwith "Invalid JSON"
```

Basically it lets you parse JSON with idiomatic F# tools like pattern matching, exceptions, and `Result`

It is a bit slow, because it has to create the ADT wrappers. You could create a different JSON parsing library which uses dynamic casts. It would be faster, but it would lose the safety of pattern matching.

**Comment by MangelMaxime** [comment]
Thanks @pauan for the explanations 👍

**Comment by alfonsogarciacaro** [comment]
Awesome work! Is it ready to merge?

**Comment by MangelMaxime** [comment]
I just want to update the JSON documentation before. As the explanation of Pauan are better than mine.

@alfonsogarciacaro I will merge tomorrow as you seem ok with the PR :)
